Add support for passing configuration to custom AccessTokenProvider

tom-s-powell opened the pull request:
A replacement for https://github.com/GoogleCloudDataproc/spark-bigquery-connector/pull/811. The goal here is to let a custom `AccessTokenProvider` be passed optional configuration via read options.

davidrabinowitz commented on bigquery-connector-common/src/test/java/com/google/cloud/bigquery/connector/common/integration/CustomCredentialsIntegrationTest.java:
```diff
@@ -60,4 +62,32 @@ public void testAccessTokenProvider() throws Exception {
+        (DefaultCredentialsDelegateAccessTokenProvider)
+            ((AccessTokenProviderCredentials) credentials).getAccessTokenProvider();
+    assertThat(accessTokenProvider.getCallCount()).isEqualTo(0);
+    assertThat(accessTokenProvider.getAccessToken().getTokenValue()).isEqualTo("some-static-token");
```
The test is not giving a valid access token, but a dummy string. This will cause the call to the BigQuery API to fail. I suggest that the field name to be "config" and to check that its value was properly set. The `getAccessToken()` implementation can remain the same.

davidrabinowitz commented:
/gcbrun

davidrabinowitz commented:
Please run `./mvnw spotless:apply -Pall` before pushing the next commit.

tom-s-powell commented on README.md:
```diff
@@ -1173,7 +1173,9 @@ spark.read.format("bigquery").option("credentials", "<SERVICE_ACCOUNT_JSON_IN_BA
   [com.google.cloud.bigquery.connector.common.AccessTokenProvider](https://github.com/GoogleCloudDataproc/spark-bigquery-connector/tree/master/bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/AccessTokenProvider.java)
   interface. The fully qualified class name of the implementation should be provided in the `gcpAccessTokenProvider`
   option. `AccessTokenProvider` must be implemented in Java or other JVM language such as Scala or Kotlin. It must
```
Ah hadn't realised this. Fixed that up.

tom-s-powell commented:
Appreciate the review @davidrabinowitz. I've addressed those comments, hope the test now looks as you were expecting? I've run `./mvnw spotless:apply -Pall` so hopefully looks okay.

davidrabinowitz commented:
/gcbrun

davidrabinowitz commented on bigquery-connector-common/src/test/java/com/google/cloud/bigquery/connector/common/integration/CustomCredentialsIntegrationTest.java:
```diff
@@ -60,4 +61,32 @@ public void testAccessTokenProvider() throws Exception {
+    // first call
+    Table table = bigQuery.getTable(TABLE_ID);
+    assertThat(table).isNotNull();
+    assertThat(accessTokenProvider.getCallCount()).isEqualTo(0);
```
Expected value is 1, as `getAccessToken()` has been called once in order to make the call the BigQuery API

tom-s-powell replied:
Fixed that

davidrabinowitz commented:
/gcbrun

tom-s-powell commented:
@davidrabinowitz couldn't quite tell, but the failing tests don't seem to be related to this change?

davidrabinowitz commented:
> @davidrabinowitz couldn't quite tell, but the failing tests don't seem to be related to this change?

Yes, we have a flaky test in the past few days. I've `@Ignore`d it in PR #917, but that's waiting for review. I'll keep an eye on this and will merge once the test passes

tom-s-powell commented:
Great thanks very much.
